storage/layers.go

```diff
@@ -2481,62 +2629,101 @@ func (r *layerStore) applyDiffWithOptions(to string, layerOptions *LayerOptions,
 	}
 	if uncompressedDigester != nil {
-		uncompressedDigest = uncompressedDigester.Digest()
+		result.uncompressedDigest = uncompressedDigester.Digest()
 	}
-	if uncompressedDigest == "" && compression == archive.Uncompressed {
-		uncompressedDigest = compressedDigest
+	if result.uncompressedDigest == "" && compression == archive.Uncompressed {
+		result.uncompressedDigest = result.compressedDigest
 	}
 
-	updateDigestMap(&r.bycompressedsum, layer.CompressedDigest, compressedDigest, layer.ID)
-	layer.CompressedDigest = compressedDigest
 	if layerOptions != nil && layerOptions.OriginalDigest != "" && layerOptions.OriginalSize != nil {
-		layer.CompressedSize = *layerOptions.OriginalSize
+		result.compressedSize = *layerOptions.OriginalSize
 	} else {
-		layer.CompressedSize = compressedCounter.Count
+		result.compressedSize = compressedCounter.Count
 	}
-	updateDigestMap(&r.byuncompressedsum, layer.UncompressedDigest, uncompressedDigest, layer.ID)
-	layer.UncompressedDigest = uncompressedDigest
-	layer.UncompressedSize = uncompressedCounter.Count
-	layer.CompressionType = compression
-	layer.UIDs = make([]uint32, 0, len(uidLog))
+	result.uncompressedSize = uncompressedCounter.Count
+	result.compressionType = compression
+
+	result.uids = make([]uint32, 0, len(uidLog))
 	for uid := range uidLog {
-		layer.UIDs = append(layer.UIDs, uid)
+		result.uids = append(result.uids, uid)
 	}
-	slices.Sort(layer.UIDs)
-	layer.GIDs = make([]uint32, 0, len(gidLog))
+	slices.Sort(result.uids)
+	result.gids = make([]uint32, 0, len(gidLog))
 	for gid := range gidLog {
-		layer.GIDs = append(layer.GIDs, gid)
+		result.gids = append(result.gids, gid)
 	}
-	slices.Sort(layer.GIDs)
+	slices.Sort(result.gids)
+
+	result.size = size
+
+	return &result, err
+}
+
+// Requires startWriting.
+func (r *layerStore) applyDiffWithOptions(to string, layerOptions *LayerOptions, diff io.Reader) (int64, error) {
+	if !r.lockfile.IsReadWrite() {
+		return -1, fmt.Errorf("not allowed to modify layer contents at %q: %w", r.layerdir, ErrStoreIsReadOnly)
+	}
+
+	layer, ok := r.lookup(to)
+	if !ok {
+		return -1, ErrLayerUnknown
+	}
+
+	tarSplitFile, err := createTarSplitFile(r, layer.ID)
+	if err != nil {
+		return -1, err
+	}
+	defer tarSplitFile.Close()
+
+	result, err := applyDiff(layerOptions, diff, tarSplitFile, func(payload io.Reader) (int64, error) {
+		options := drivers.ApplyDiffOpts{
+			Diff:       payload,
+			Mappings:   r.layerMappings(layer),
+			MountLabel: layer.MountLabel,
+		}
+		return r.driver.ApplyDiff(layer.ID, options)
+	})
+	if err != nil {
+		return -1, err
+	}
+
+	if err := tarSplitFile.Sync(); err != nil {
+		return -1, fmt.Errorf("sync tar-split file: %w", err)
+	}
+
+	applyDiffResultToLayer(r, layer, layerOptions, result)
 
 	err = r.saveFor(layer)
 
-	return size, err
+	return result.size, err
+}
+
+func applyDiffResultToLayer(r *layerStore, layer *Layer, layerOptions *LayerOptions, result *applyDiffResult) {
```

Comment:
```suggestion
// Requires startWriting.
func applyDiffResultToLayer(r *layerStore, layer *Layer, layerOptions *LayerOptions, result *applyDiffResult) {
```
and perhaps make it a method on `layerStore`, when it is updating its state fields.